packages/markerikson-stack-client-prototype/src/features/sources/sourcesCache.ts

```diff
@@ -0,0 +1,118 @@
+    a consistent promise reference if the data isn't yet available.  
+    (Note: the promise throwing behavior may still be subject to change? per Dan, 
+      https://github.com/facebook/react/issues/22964#issuecomment-1113649154 )
+
+  Brian's example uses synchronous "wakeables" - a plain JS object
+  with a `.then` method, plus `.resolve/reject` methods, and is doing
+  manual "pending/resolved" tracking.  This is apparently to avoid the
+  overhead of promises and microtasks.
+
+  RTKQ recently added a `getRunningOperationPromise` method, which returns
+  a consistent promise reference _if_ there's a request in flight for that
+  particular endpoint+args combination.  So, we can use that as the thrown promise.
+
+  I've copy-pasted some of the boolean flag derivation logic from the guts of
+  RTKQ to help figure out if the RTKQ cache entry has data available or not.
+ */
+
+let store: AppStore;
+
+export const setStore = (_store: AppStore) => {
+  store = _store;
+};
+
+// Placeholder for the result given to React's Suspense Cache
+interface SourceGroupsResult {
+  groups: SourceGroups | null;
+}
+
+// We know there's only one cache entry, since it's a "list of things" query.
+// Create a selector to read that cache entry from the Redux store.
+const selectSourceGroupsEntry = api.endpoints.getSources.select();
+
+function createSourceGroupsRecord(): SourceGroupsResult {
+  return {
+    groups: null,
+  };
+}
+
+export function getSourceGroups(): SourceGroups {
+  // Create or read the cache for this section of the UI (????)
+  const sourceGroupsRecord = getCacheForType(createSourceGroupsRecord);
+
+  // Do we have this data already cached? If so, return it
+  if (sourceGroupsRecord.groups !== null) {
+    console.log("Found existing cached sources data, returning");
+    return sourceGroupsRecord.groups;
+  }
+
+  // RTKQ side: have we ever tried to request this data?
+  let sourcesEntry: ReturnType<typeof selectSourceGroupsEntry> | undefined =
+    selectSourceGroupsEntry(store.getState());
+
+  if (!sourcesEntry || sourcesEntry.isUninitialized) {
+    console.log("No RTKQ entry found, initiating request");
+    // If not, actually start fetching
+    store.dispatch(api.endpoints.getSources.initiate());
+
+    // Now we've got an entry in the store
+    sourcesEntry = selectSourceGroupsEntry(store.getState());
+  }
+
+  //Copy-pasted from the guts of RTKQ's `buildHooks.ts`,
+  // since this is derived state
+
+  // data is the last known good request result we have tracked - or if none has been tracked yet the last good result for the current args
+  let data = sourcesEntry!.data;
+
+  const hasData = data !== undefined;
+
+  // isFetching = true any time a request is in flight
+  const isFetching = sourcesEntry!.isLoading;
+  // isLoading = true only when loading while no data is present yet (initial load with no data in the cache)
+  const isLoading = !hasData && isFetching;
```

Comment:
It's not clear to me why `sourcesEntry!.isLoading` isn't sufficient for this. RTK should reset that field to `false` once it has data, no?

Reply:
RTKQ differentiates between "loading", which is only true during the initial setup, and "fetching", which can go between false/true many times if you change the parameters being passed into a given query hook:

In this case, we're doing the work outside of any hook, so I actually copy-pasted this logic from inside of RTKQ's hooks code.  It may not be totally necessary here, I'd have to think about it a bit more.